cedar-policy/src/tests.rs

```diff
@@ -0,0 +1,2737 @@
+                .expect("lookup failed")
+                .id(),
+            &"template".parse().unwrap()
+        );
+        assert_eq!(
+            pset.policy(&"static".parse().unwrap())
+                .expect("lookup failed")
+                .id(),
+            &"static".parse().unwrap()
+        );
+        assert_eq!(
+            pset.policy(&"linked".parse().unwrap())
+                .expect("lookup failed")
+                .id(),
+            &"linked".parse().unwrap()
+        );
+    }
+
+    #[test]
+    fn link_static_policy() {
+        // Linking the `PolicyId` of a static policy should not be allowed.
+        // Attempting it should cause an `ExpectedTemplate` error.
+        let static_policy = Policy::parse(
+            Some("static".into()),
+            "permit(principal, action, resource);",
+        )
+        .expect("Static parse failure");
+        let mut pset = PolicySet::new();
+        pset.add(static_policy).unwrap();
+
+        let before_link = pset.clone();
+        let result = pset.link(
+            PolicyId::from_str("static").unwrap(),
+            PolicyId::from_str("linked").unwrap(),
+            HashMap::new(),
+        );
+        assert_matches!(result, Err(PolicySetError::ExpectedTemplate));
+        assert_eq!(
+            pset, before_link,
+            "A failed link shouldn't mutate the policy set"
+        );
+    }
+
+    #[test]
+    fn link_linked_policy() {
+        let template = Template::parse(
+            Some("template".into()),
+            "permit(principal == ?principal, action, resource);",
+        )
+        .expect("Template Parse Failure");
+        let mut pset = PolicySet::new();
+        pset.add_template(template).unwrap();
+
+        pset.link(
+            PolicyId::from_str("template").unwrap(),
+            PolicyId::from_str("linked").unwrap(),
+            std::iter::once((SlotId::principal(), EntityUid::from_strs("Test", "test"))).collect(),
+        )
+        .unwrap();
+
+        let before_link = pset.clone();
+        let result = pset.link(
+            PolicyId::from_str("linked").unwrap(),
+            PolicyId::from_str("linked2").unwrap(),
+            HashMap::new(),
+        );
+        assert_matches!(result, Err(PolicySetError::ExpectedTemplate));
+        assert_eq!(
+            pset, before_link,
+            "A failed link shouldn't mutate the policy set"
+        );
+    }
+
+    #[cfg(feature = "partial-eval")]
+    #[test]
+    fn unknown_entities() {
+        let ast = ast::Policy::from_when_clause(
+            ast::Effect::Permit,
+            ast::Expr::unknown(ast::Unknown::new_with_type(
+                "test_entity_type::\"unknown\"",
+                ast::Type::Entity {
+                    ty: ast::EntityType::Concrete("test_entity_type".parse().unwrap()),
+                },
+            )),
+            ast::PolicyID::from_smolstr("static".into()),
+        );
+        let static_policy = Policy::from_ast(ast);
```

Comment:
Needed for this.

It's unclear to me if we want people to be able to directly construct policies with unknown entities (i.e., do we want `Policy::from_ast` to be `pub` with `partial-eval` enabled). OTOH, I don't know why we enforce this property in a unit test if we don't intend to allow people to construct policies with unknown entities from our public API. If we don't intend that, I think we should test that the output of some PE has the expected unknown entities.

Reply:
Not familiar with this test, but if the test requires `ast::Policy`, then perhaps it shouldn't be in the `cedar-policy` crate?